text/0000-cargo-report-future-incompat.md

```diff
@@ -0,0 +1,720 @@
+- Feature Name: N/A
+- Start Date: 2019-12-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: N/A
+
+# Summary
+[summary]: #summary
+
+Cargo should alert developers to upstream dependencies that trigger
+future-incompatibility warnings. Cargo should list such dependencies
+even when these warnings have been suppressed (e.g. via cap-lints or
+`#[allow(..)]` attributes.)
+
+Cargo could additionally provide feedback for tactics a maintainer of
+the downstream crate could use to address the problem (the details of
+such tactics is not specified nor mandated by this RFC).
+
+# Motivation
+[motivation]: #motivation
+
+From [rust-lang/rust#34596][]:
+
+> if you author a library that is widely used, but which you are not
+> actively using at the moment, you might not notice that it will
+> break in the future -- moreover, your users won't either, since
+> cargo will cap lints when it builds your library as a dependency.
+
+[rust-lang/rust#34596]: https://github.com/rust-lang/rust/issues/34596
+
+Today, cargo will cap lints when it builds libraries as dependencies.
+This behavior includes future-incompatibility lints.
+
+As a running example, assume we have a crate `unwary` with an upstream
+crate dependency `brash`, and `brash` has code that triggers a
+future-incompatibility lint, in this case a borrow `&x.data.0` of a packed field
+(see [rust-lang/rust#46043][], "safe packed borrows").
+
+[rust-lang/rust#46043]: https://github.com/rust-lang/rust/issues/46043
+
+If `brash` is a non-path dependency of `unwary`, then building
+`unwary` will suppress the warning associated with `brash` in its
+diagnostic output, because the build of `brash` will pass
+`--cap-lints=allow` to its `rustc` invocation. This means that a
+future version of Rust is going to fail to compile the `unwary`
+project, with no warning to the developer of `unwary`.
+
+Example of today's behavior (where in this case, `brash` is non-path dependency of `unwary`):
```

Comment:
Since a few weeks Cargo does cache warnings and displays them during every build.

Reply:
okay I'll update this.

Reply:
(I assume this note was meant to be attached to the digression regarding behavior on path dependencies. The behavior on non-path dependencies remains, as one would expect.)

Reply:
I removed the aforementioned digression regarding path dependencies in 4126f2d7c926e4cab15c53d5df34e03ee99ffe31 .